### PR TITLE
ambient: ca: Implement node authorization

### PR DESCRIPTION
**Pull request by howardjohn**
This implements the node authorization described in https://github.com/istio/istio/issues/41281

Note: this only implements the CA part. The ztunnel and XDS will be updated in a later PR. Because this is extremely security sensitive code, I wanted to keep the PR as small as possible.

**Please provide a description of this PR:**

### CODE REVIEW

File: security/pkg/server/ca/node_auth.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		// This should never happen (only possible if the index doesn't exist)
+		return fmt.Errorf("failed to find node local service accounts: %v", err)
+	}
+	// We don't care what pods are part of the index, only that there is at least one. If there is one,
+	// it is appropriate for the caller to request this identity.
+	if len(res) == 0 {
+		return fmt.Errorf("no instances of %q found on node %q", k.ServiceAccount, k.Node)
+	}
+	serverCaLog.Debugf("Node caller ")
```
Comment (lei-tang) [inline security/pkg/server/ca/node_auth.go:114]: > serverCaLog.Debugf("Node caller ")
Should the Debugf statement here include who is the node caller?

File: security/pkg/server/ca/node_auth.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		Node:           callerPod.Spec.NodeName,
+	}
+	// TODO: this is currently single cluster; we will need to take the cluster of the proxy into account
+	// to support multi-cluster properly.
+	res, err := na.podIndexer.ByIndex(NodeSaIndex, k.String())
+	if err != nil {
+		// This should never happen (only possible if the index doesn't exist)
+		return fmt.Errorf("failed to find node local service accounts: %v", err)
+	}
+	// We don't care what pods are part of the index, only that there is at least one. If there is one,
```
Comment (lei-tang) [inline security/pkg/server/ca/node_auth.go:109]: Does this function checks that the ztunnel and the pod for which the ztunnel requests a certificate are on the same node?

File: security/pkg/server/ca/node_auth.go (hunk 1 above)
Comment (howardjohn) [reply security/pkg/server/ca/node_auth.go:114]: Yes! Good catch... I must have got distracted half way through writing this line :-)

File: security/pkg/server/ca/node_auth.go (hunk 2 above)
Comment (howardjohn) [reply security/pkg/server/ca/node_auth.go:109]: Exactly

**Review by lei-tang — APPROVED** [review]

**Review by costinm — COMMENTED** [review]
Right now we don't use the common name part of the cert - any chance we can add something there (not sure in what format) to support a way to identify such certificates ? It may also 
be worth adding the node ID and pod ID - may be a bit of an abuse of the CN field, not sure if there is a better place to convey this information.

File: pilot/pkg/features/pilot.go [hunk 3]
```diff
@@ -418,6 +419,30 @@ var (
 		return strings.Split(trustedGatewayCIDR.Get(), ",")
 	}()
 
+	CATrustedNodeAccounts = func() map[types.NamespacedName]struct{} {
```
Comment (stevenctl) [inline pilot/pkg/features/pilot.go:422]: How permanent is this intended to be? If we want to stop trusting some service account, we'd have to restart istiods.

**Review by stevenctl — APPROVED** [review]

File: pilot/pkg/features/pilot.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/features/pilot.go:422]: Intent is you have 1 ztunnel and that is the only one ever configured. We can move to mesh config if needed though

Comment (stevenctl) [reply pilot/pkg/features/pilot.go:422]: The maybe impossible scenario I imagined is somehow something talking to istiod manages to hijack the identity of the ztunnel and grab all the node certs.

**Pull request merged**